### PR TITLE
Add distance constraint to DIRCON

### PR DESCRIPTION
**Pull request by mposa**
-Fixes an issue with MultibodyUtils when a plant has no actuators
-Add DirconDistanceData, to represent the Cassie linkage within DIRCON.

### CODE REVIEW

File: systems/trajectory_optimization/dircon_distance_data.cc [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+  // Jdot*v can then be computed as
+  //    2*(r1dot-r2dot)^T * (J1 - J2)*v + 2*(r1-r2)^T * (J1dot - J2dot)*v
+  //  = 2*(J1*v - J2*v)^T * (J1 - J2)*v + 2*(r1-r2)^T * (J1dot - J2dot)*v
+  //  = 2*||(J1-J2)*v||^2 + 2*(r1-r2)^T * (J1dot - J2dot)*v
+  // And cdot = J*v (as usual)
+  VectorX<T> pt_delta = pt1_transform - pt2_transform;
+  this->c_(0) = pt_delta.squaredNorm() - distance_ * distance_;
+  this->J_ = 2*pt_delta.transpose()*(J1 - J2);
+  this->Jdotv_ = 2*pt_delta.transpose()*(J1_times_v - J2dot_times_v);
+  this->Jdotv_(0) += (this->J_*pt_delta).squaredNorm();
```
Comment (yminchen) [inline systems/trajectory_optimization/dircon_distance_data.cc]: Should we replace `pt_delta` with `v`?

Comment (mposa) [reply systems/trajectory_optimization/dircon_distance_data.cc]: Good catch. It's also missing a 2.

File: systems/trajectory_optimization/dircon_distance_data.cc [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+  // Jdot*v can then be computed as
+  //    2*(r1dot-r2dot)^T * (J1 - J2)*v + 2*(r1-r2)^T * (J1dot - J2dot)*v
+  //  = 2*(J1*v - J2*v)^T * (J1 - J2)*v + 2*(r1-r2)^T * (J1dot - J2dot)*v
+  //  = 2*||(J1-J2)*v||^2 + 2*(r1-r2)^T * (J1dot - J2dot)*v
+  // And cdot = J*v (as usual)
+  VectorX<T> pt_delta = pt1_transform - pt2_transform;
+  this->c_(0) = pt_delta.squaredNorm() - distance_ * distance_;
+  this->J_ = 2*pt_delta.transpose()*(J1 - J2);
+  this->Jdotv_ = 2*pt_delta.transpose()*(J1_times_v - J2dot_times_v);
+  this->Jdotv_(0) += 2*(((J1 - J2)*v)).squaredNorm();
```
Comment (yminchen) [inline systems/trajectory_optimization/dircon_distance_data.cc]: Similarly, should square this after taking a square norm. 
E.g. `this->Jdotv_(0) += 2*pow(((J1 - J2)*v).squaredNorm(),2);`

File: systems/trajectory_optimization/dircon_distance_data.cc [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+  // Constraint is ||r1-r2||^2  - d^2, to keep it differentiable everywhere
+  // J is then 2*(r1-r2)^T * (J1 - J2)
+  // Jdot*v can then be computed as
+  //    2*(r1dot-r2dot)^T * (J1 - J2)*v + 2*(r1-r2)^T * (J1dot - J2dot)*v
+  //  = 2*(J1*v - J2*v)^T * (J1 - J2)*v + 2*(r1-r2)^T * (J1dot - J2dot)*v
+  //  = 2*||(J1-J2)*v||^2 + 2*(r1-r2)^T * (J1dot - J2dot)*v
+  // And cdot = J*v (as usual)
+  VectorX<T> pt_delta = pt1_transform - pt2_transform;
+  this->c_(0) = pt_delta.squaredNorm() - distance_ * distance_;
```
Comment (yminchen) [inline systems/trajectory_optimization/dircon_distance_data.cc]: Need to square `pt_delta.squaredNorm()`

File: systems/trajectory_optimization/dircon_distance_data.cc (hunk 1 above)
Comment (yminchen) [reply systems/trajectory_optimization/dircon_distance_data.cc]: Double checked and added new comments

**Comment by mposa** [comment]
"Squarednorm" is norm squared (no need to square it)

**Comment by yminchen** [comment]
> "Squarednorm" is norm squared (no need to square it) 

Ah, you are right.

**Pull request merged**